### PR TITLE
[BATCH] clean sensitivity analysis method in exploration

### PR DESCRIPTION
**Pull request by chapuisk**
Sensitivity analysis methods, have been added to Gama as of 1.9.x, but have never been polished since, notably in the provided feedbacks and outputs. I list the foreseen improvement 

- [x] Morris : alternative report in the form of a proper csv for external visualization (R, python, etc.) 
- [x] beta_d : Same alternative report 
- [x] Proper doc for stochastic analysis

There is also part of the improvement to further deploy SA features to every exploration methods:

- [x] All exploration methods should be able to write raw simulation results, using optional ``outputs`` and ``results`` facets. See #147

Don't hesitate to add more task to upgrade SA & exploration in Gama.

### CODE REVIEW

**Comment by lesquoyb** [comment]
Ok so I've read your remaining spotbugs violations, and I made a more human readable summary with my point of view on how seriously you should take them:

1. the 3 CT_CONSTRUCTOR_THROW it's up to you to fix it or not but I don't think it's important, if an exception is raised in those constructors gama will probably catch it and display it the user so there's no real risk even though it's a bit ugly
2. the 6 WMI_WRONG_MAP_ITERATOR are just telling you that you could use `entrySet` to get the key and value in the same object for your loops instead of iterating on `keySet` and get the value with the `get` function. To me it doesn't matter if you fix it or not as it's not in parts of the code where optimization could be important.
3. the 2 RV_RETURN_VALUE_IGNORED_BAD_PRACTICE are more important as they highlight that you didn't handle the case of error while creating a directory or deleting a file. I think you should probably use the `GAMA.reportAndThrowIfNeeded` method to stop the execution and report an error.
4. the SBSC_USE_STRINGBUFFER_CONCATENATION indicates that you could have used the `StringBuffer`'s `append` method instead of concatenated strings, which is more efficient, and particularly in a loop it could amount to good performance gains. I advise that you do it.
5. the 3 EI_EXPOSE_REP2 you could probably ignore. It indicates that you are keeping a reference on lists/maps that could be modified from the outside. If you want to be "clean" you could probably clone those lists/maps
6. the NP_ALWAYS_NULL is **very important**, it indicates that you are trying to access a method on an object that we already know is null, thus it will raise a `NullPointerException`. Reading the code it seems like the condition is wrong (`&&` instead of `||`) but as you told me you never have any exception you should consider completely deleting this constructor as it's probably never called
7. the NP_NULL_PARAM_DEREF is raised because it is (theoretically) possible to reach this code with a value of `null` for the variable `fr` which will lead to more exception. In reality it may not be reached because some exceptions are raised before, but I'm not sure. I recommend that either you explicitly use a return in the catch blocks to make sure that this code is never reach in case of exception, or you move all the code following the initialization of `fr` in the `try` block.
8. the OS_OPEN_STREAM is raised because you open a stream but it is never explicitly closed in case of exception. I think you should also solve this one and this can be done using a try with resource. It would be something like: `try (BufferedReader br = new BufferedReader(fr)){...`. This construction will make sure that br is disposed (and thus closed) at the end of the block.
9. And finally there are 3 RV_EXCEPTION_NOT_THROWN and I think those are **really important** as it means you have created exceptions that are not thrown. I've also recently learned about this, but `GamaRuntimeException.error` only creates a `GamaRuntimeException` object, and you still need to raise the exception for it to be seen by the user. You can do so either with `throw GamaRuntimeException.create...` or using the method `GAMA.reportAndThrowIfNeeded`. I personally use the `GAMA.reportAndThrowIfNeeded` syntax as I think it's the most `gama` way but I'm not 100% sure it's the best option every time.

To be honest I could have resolved almost all of those by myself and commited the fixes, but I think it's important to do that kind of explanation as you will probably encounter them again and I might not always be available to fix them.

**Comment by lesquoyb** [comment]
Some additional information: I found out that the compilation check that fails is actually due to unit tests, that's why when I tried your branch on my computer I didn't have any problem, but if I try to run the unit tests it is indeed broken and I can track some of the errors stacks to `AExplorationAlgorithm`:
```
java.lang.ClassCastException: class gama.core.kernel.experiment.TestAgent cannot be cast to class gama.core.kernel.simulation.SimulationAgent (gama.core.kernel.experiment.TestAgent and gama.core.kernel.simulation.SimulationAgent are in unnamed module of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @76680c63)
	at gaml.additions.core.GamlAdditions.lambda$146(GamlAdditions.java:232)
	at gama.gaml.compilation.GamaHelper.run(GamaHelper.java:75)
	at gama.gaml.compilation.IGamaHelper.run(IGamaHelper.java:51)
	at gama.gaml.variables.Variable.value(Variable.java:822)
	at gama.gaml.variables.Variable.value(Variable.java:816)
	at gama.gaml.variables.Variable.getInitialValue(Variable.java:871)
	at gama.core.kernel.experiment.ExperimentAgent$ExperimentAgentScope.getGlobalVarValue(ExperimentAgent.java:1100)
	at gama.gaml.expressions.variables.GlobalVariableExpression._value(GlobalVariableExpression.java:95)
	at gama.gaml.expressions.AbstractExpression.value(AbstractExpression.java:87)
	at gama.gaml.expressions.operators.BinaryOperator._value(BinaryOperator.java:148)
	at gama.gaml.expressions.AbstractExpression.value(AbstractExpression.java:87)
	at gama.gaml.expressions.operators.BinaryOperator._value(BinaryOperator.java:149)
	at gama.gaml.expressions.AbstractExpression.value(AbstractExpression.java:87)
	at gama.gaml.variables.Variable.getInitialValue(Variable.java:866)
	at gama.core.kernel.experiment.ExperimentAgent$ExperimentAgentScope.getGlobalVarValue(ExperimentAgent.java:1100)
	at gama.gaml.expressions.variables.GlobalVariableExpression._value(GlobalVariableExpression.java:95)
	at gama.gaml.expressions.AbstractExpression.value(AbstractExpression.java:87)
	at gama.gaml.expressions.operators.UnaryOperator._value(UnaryOperator.java:97)
	at gama.gaml.expressions.AbstractExpression.value(AbstractExpression.java:87)
	at gama.gaml.expressions.operators.BinaryOperator._value(BinaryOperator.java:148)
	at gama.gaml.expressions.AbstractExpression.value(AbstractExpression.java:87)
	at gama.gaml.expressions.operators.BinaryOperator._value(BinaryOperator.java:148)
	at gama.gaml.expressions.AbstractExpression.value(AbstractExpression.java:87)
	at gama.gaml.expressions.operators.BinaryOperator._value(BinaryOperator.java:149)
	at gama.gaml.expressions.AbstractExpression.value(AbstractExpression.java:87)
	at gama.gaml.expressions.operators.BinaryOperator._value(BinaryOperator.java:148)
	at gama.gaml.expressions.AbstractExpression.value(AbstractExpression.java:87)
	at gama.gaml.statements.test.AssertStatement.privateExecuteIn(AssertStatement.java:100)
	at gama.gaml.statements.AbstractStatement.executeOn(AbstractStatement.java:47)
	at gama.gaml.statements.test.TestStatement.privateExecuteIn(TestStatement.java:145)
	at gama.gaml.statements.AbstractStatement.executeOn(AbstractStatement.java:47)
	at gama.gaml.statements.AbstractStatementSequence.executeOn(AbstractStatementSequence.java:59)
	at gama.core.runtime.ExecutionScope.execute(ExecutionScope.java:510)
	at gama.core.runtime.IScope.execute(IScope.java:443)
	at gama.core.runtime.IScope.execute(IScope.java:414)
	at gama.gaml.architecture.reflex.ReflexArchitecture.executeReflexes(ReflexArchitecture.java:126)
	at gama.gaml.architecture.reflex.ReflexArchitecture.executeOn(ReflexArchitecture.java:112)
	at gama.core.runtime.ExecutionScope.execute(ExecutionScope.java:510)
	at gama.core.runtime.IScope.execute(IScope.java:443)
	at gama.core.metamodel.agent.AbstractAgent.doStep(AbstractAgent.java:269)
	at gama.core.metamodel.agent.MinimalAgent.doStep(MinimalAgent.java:260)
	at gama.core.metamodel.agent.AbstractAgent.step(AbstractAgent.java:240)
	at gama.core.kernel.experiment.BatchAgent.launchSimulationsWithSolution(BatchAgent.java:368)
	at gama.core.kernel.batch.exploration.Exploration.explore(Exploration.java:194)
	at gama.core.kernel.batch.exploration.AExplorationAlgorithm.run(AExplorationAlgorithm.java:124)
	at gama.core.kernel.experiment.BatchAgent.step(BatchAgent.java:233)
	at gama.core.kernel.experiment.TestAgent.step(TestAgent.java:93)
	at gaml.compiler.ui.utils.ModelRunner.runHeadlessTests(ModelRunner.java:114)
	at gama.ui.shared.utils.SwtGui.runHeadlessTests(SwtGui.java:391)
	at gama.ui.shared.commands.TestsRunner.lambda$0(TestsRunner.java:69)
	at org.eclipse.core.runtime.jobs.Job$2.run(Job.java:187)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
```
So at least that "mystery" is solved

File: gama.core/src/gama/core/kernel/batch/exploration/stochanalysis/StochanalysisExploration.java [hunk 1]
```diff
@@ -186,7 +191,7 @@ public void explore(final IScope scope) throws GamaRuntimeException {
 				hasFacet(IKeyword.BATCH_OUTPUT)) { saveRawResults(scope, res_outputs); }
 		
 		/** If any of the two facet is missing pop up a warning */
-		if (hasFacet(IKeyword.BATCH_VAR_OUTPUTS) ||
+		if (hasFacet(IKeyword.BATCH_VAR_OUTPUTS) &&
```
Comment (lesquoyb) [inline gama.core/src/gama/core/kernel/batch/exploration/stochanalysis/StochanalysisExploration.java:194]: Shouldn't it be the opposite: ` ! (hasFacet(...) && hasFacet(...) )` ?
With the current code you will print a warning in the case both facets are present

**Comment by lesquoyb** [comment]
Hello @chapuisk 

I've committed a few changes to address some of the spotbugs warnings that were "easy" to fix.
There are 10 remaining, out of those I think we can ignore the `WMI_WRONG_MAP_ITERATOR ` and the 3 `EI_EXPOSE_REP2 `.

So that leaves us with 2 `RV_RETURN_VALUE_IGNORED_BAD_PRACTICE` and 3 `CT_CONSTRUCTOR_THROW`, those are up to you to fix or not as it's more a question of design: do you want to let the execution continue in case there's an issue with deleting/creating a directory with the risk that it will cause troubles later or is it safe enough? And for the constructors, the "risk" mentioned by spotbugs doesn't really apply to us so we could ignore, but at the same time it's a pretty ugly design so unless there's no other choice I think it's always better not to throw exception in constructors.

In addition to that, there's one remaining `RV_EXCEPTION_NOT_THROWN`. While it actually doesn't matter, it is in a part of the code that you left unfinished in MorrisExploration. Will this code be completed in that PR or is it something you want to leave for latter ?

There's also this potential logical error I mentioned in a comment above (see [this](https://github.com/gama-platform/gama/pull/186#discussion_r1665309691). The fix I suggested may not be the right one depending on when exactly you want to trigger a warning, but I'm quite sure the current code is wrong.

**Comment by chapuisk** [comment]
@lesquoyb code smell is globally better after PR, but still not above the threshold of SpotBugs. However, I think merge PR is ok for me. Do you agree?

**Comment by lesquoyb** [comment]
Yes, all good for me, I've also added two more controls over the file manipulation function calls, now it's time to break gama!

**Pull request merged**